views/js/runner/navigator/navigator.js

```diff
@@ -75,91 +47,8 @@ define([
-            isLeavingTestPart = (newTestPart.id !== testContext.testPartId);
-
-            //guess the new testContext data
-            newContext = _.defaults({
-                itemIdentifier : newItem.id,
-                itemPosition   : position,
-
-                // when the test part is linear, the item is always answered as we cannot come back to it
-                itemAnswered   : newItem.answered || newTestPart.isLinear,
-
-                //FIXME numberPresented can be late
-                numberPresented : testMap.stats.viewed,
-                numberCompleted : testMap.stats.answered,
-
-                //FIXME unsupported yet
-                hasFeedbacks : false,
-
-                //FIXME maintain attempts
-                //FIXME attempts can be incorrect (based on last known value)
-                remainingAttempts : (newItem.remainingAttempts > -1) ? newItem.remainingAttempts - 1 : -1,
-
-                sectionId:       newSection.id,
-                sectionTitle:    newSection.label,
-                testPartId:      newTestPart.id,
-                isLinear:        newTestPart.isLinear,
-                isLast:          navigationHelper.isLast(testMap, newItem.id),
-                canMoveBackward: !newTestPart.isLinear && !navigationHelper.isFirst(testMap, newItem.id)
-            }, testContext);
-
-            //if the section is different, we don't keep the rubric blocks
-            if(isLeavingSection){
-                newContext.numberRubrics = 0;
-                newContext.rubrics = '';
-            }
-
-            //remove timers if they're not on the same scope
-            newContext.timeConstraints = _.reject(testContext.timeConstraints, function(constraint){
-                return constraint.qtiClassName === 'assessmentItemRef' ||
-                       (isLeavingSection && constraint.qtiClassName === 'assessmentSection') ||
-                       (isLeavingTestPart && constraint.qtiClassName === 'testPart');
-            });
-
-            if(newItem.timeConstraint){
-                newContext.timeConstraints.push(newItem.timeConstraint);
-            }
-            if(isLeavingSection && newSection.timeConstraint){
-                newContext.timeConstraints.push(newSection.timeConstraint);
-            }
-            if(isLeavingTestPart && newTestPart.timeConstraint){
-                newContext.timeConstraints.push(newTestPart.timeConstraint);
-            }
-
-            //FIXME default options may not be accurate
-            newContext.options = _.defaults(
-                getOptionsFromCategories(newItem.categories || []),
-                _.pick(testContext.options, ['allowComment', 'allowSkipping', 'exitButton', 'logoutButton'])
-            );
-
-            return newContext;
-        };
+    var navigatorFactory = function navigatorFactory(testData, testContext, testMap) {
+        var testContextBuilder = new TestContextBuilder(testData, testContext, testMap);
```

Comment:
Caution you are using a factory as constructor.

Isn't the `TestContextBuilder` a helper? if so should not it be a static object instead? Having to instanciate a helper could lead to issues.